webrtc/dtls/dtls_transport.nim

```diff
@@ -0,0 +1,179 @@
+
+proc new*(T: type Dtls, transport: Stun): T =
+  var self = T(
+    connections: initTable[TransportAddress, DtlsConn](),
+    transport: transport,
+    laddr: transport.laddr,
+    started: true
+  )
+
+  mb_ctr_drbg_init(self.ctr_drbg)
+  mb_entropy_init(self.entropy)
+  mb_ctr_drbg_seed(self.ctr_drbg, mbedtls_entropy_func, self.entropy, nil, 0)
+
+  self.serverPrivKey = self.ctr_drbg.generateKey()
+  self.serverCert = self.ctr_drbg.generateCertificate(self.serverPrivKey)
+  self.localCert = newSeq[byte](self.serverCert.raw.len)
+  copyMem(addr self.localCert[0], self.serverCert.raw.p, self.serverCert.raw.len)
+  return self
+
+proc stop*(self: Dtls) {.async.} =
+  if not self.started:
+    warn "Already stopped"
+    return
+
+  await allFutures(toSeq(self.connections.values()).mapIt(it.close()))
+  self.started = false
+
+# -- Remote / Local certificate getter --
+
+proc localCertificate*(self: Dtls): seq[byte] =
+  self.localCert
+
+# -- Dtls Accept / Connect procedures --
+
+proc cleanupDtlsConn(self: Dtls, conn: DtlsConn) {.async.} =
+  # Waiting for a connection to be closed to remove it from the table
+  await conn.join()
+  self.connections.del(conn.raddr)
+
+proc accept*(self: Dtls): Future[DtlsConn] {.async.} =
+  ## Accept a Dtls Connection
+  ##
+  var res = DtlsConn.new(await self.transport.accept(), self.laddr)
+
+  mb_ssl_init(res.ctx.ssl)
+  mb_ssl_config_init(res.ctx.config)
+  mb_ssl_cookie_init(res.ctx.cookie)
+  mb_ssl_cache_init(res.ctx.cache)
+
+  res.ctx.ctr_drbg = self.ctr_drbg
+  res.ctx.entropy = self.entropy
+
+  res.ctx.pkey = self.serverPrivKey
+  res.ctx.srvcert = self.serverCert
+  res.localCert = self.localCert
+
+  mb_ssl_config_defaults(
+    res.ctx.config,
+    MBEDTLS_SSL_IS_SERVER,
+    MBEDTLS_SSL_TRANSPORT_DATAGRAM,
+    MBEDTLS_SSL_PRESET_DEFAULT
+  )
+  mb_ssl_conf_rng(res.ctx.config, mbedtls_ctr_drbg_random, res.ctx.ctr_drbg)
+  mb_ssl_conf_read_timeout(res.ctx.config, 10000) # in milliseconds
+  mb_ssl_conf_ca_chain(res.ctx.config, res.ctx.srvcert.next, nil)
+  mb_ssl_conf_own_cert(res.ctx.config, res.ctx.srvcert, res.ctx.pkey)
+  mb_ssl_cookie_setup(res.ctx.cookie, mbedtls_ctr_drbg_random, res.ctx.ctr_drbg)
+  mb_ssl_conf_dtls_cookies(res.ctx.config, addr res.ctx.cookie)
+  mb_ssl_set_timer_cb(res.ctx.ssl, res.ctx.timer)
+  mb_ssl_setup(res.ctx.ssl, res.ctx.config)
+  mb_ssl_session_reset(res.ctx.ssl)
+  mb_ssl_set_verify(res.ctx.ssl, verify, res)
+  mb_ssl_conf_authmode(res.ctx.config, MBEDTLS_SSL_VERIFY_OPTIONAL)
+  mb_ssl_set_bio(res.ctx.ssl, cast[pointer](res), dtlsSend, dtlsRecv, nil)
+  while true:
+    try:
+      self.connections[res.raddr] = res
+      await res.dtlsHandshake(true)
+      asyncSpawn self.cleanupDtlsConn(res)
+      break
+    except WebRtcError as exc:
+      trace "Handshake fails, try accept another connection",
+            remoteAddress = res.raddr, error = exc.msg
+      self.connections.del(res.raddr)
+      res.conn = await self.transport.accept()
```

Comment:
This whole code seems a bit strange to me. It doesn't seem right that we can just replace the `conn` like that.

Reply:
You're right, there was some issue with this bit of code. I changed it.